### DISCUSSION TITLE
Cursor block style sometimes become a line

### DISCUSSION THREAD

**Discussion by yotamNimble** [ghostty-org/ghostty#3893] [Q&A] [answered]
I have a script on Fish startup to choose a Tmux session. When I click enter I get the expected behavior:
https://github.com/user-attachments/assets/f57349bf-327c-4472-9340-f4b72f041edb

When I click `ctrl+d` which I use to not open any Tmux session, the cursor looks like this:
https://github.com/user-attachments/assets/5fcd329c-fd55-45fd-8fa7-446fafaf75d7

I'm using Fish Shell version 3.7.0 on Fedora 41 with everything up-to-date.
Ghostty itself is installed from a copr and this is the version output:
```
yotam@yotams-x1-carbon ~> ghostty --version
Ghostty 1.0.0-dev+0000000

Version
  - version: 1.0.0-dev+0000000
  - channel: tip
Build Config
  - Zig version: 0.13.0
  - build mode : builtin.OptimizeMode.ReleaseFast
  - app runtime: apprt.Runtime.gtk
  - font engine: font.main.Backend.fontconfig_freetype
  - renderer   : renderer.OpenGL
  - libxev     : main.Backend.io_uring
  - GTK version:
    build      : 4.16.5
    runtime    : 4.16.5
  - libadwaita : enabled
    build      : 1.6.2
    runtime    : 1.6.2
```

**Comment by magnusriga** [comment]
I am experiencing the same issue with OrbStack.

I have set up shell-intergration, double-checked the Ghostty `config` file, and ensured terminfo is copied into the linux machine.

This is in my `config`:

```
cursor-style = block
cursor-style-blink = false
shell-integration-features = no-cursor,sudo,title
```

As configured, when first launching Ghostty on OSX, it shows a non-blinking block cursor. Then, when entering an OrbStack Linux machine from the terminal, (`orb`), the cursor becomes a blinking bar.

The same thing happens when I hit `ctrl-d` to escape out of Linux and back into the OSX environment. Meaning, the cursor has then become a blinking bar.

**PS:** I have tried re-sourcing the shell-integration script manually, which does not change anything. In fact, since the cursor is a blinking bar, it does seem like shell-integration is working as it should. The problem therefore seems to be that the Ghostty `config` is not respected inside the Linux box, or when `ctrl-d` back to the host machine.

**PS 2:** Running `ghostty +show-config` inside both the OSX terminal and the Linux orb terminal, shows that `shell-integration-features = no-cursor,sudo,title` is set. Thus, it seems odd that the cursor is reverted to a blinking bar.

**Reply by 00-kat** [reply]
From the documentation for [`cursor-style`]:

> A running program can still request an explicit cursor style using escape sequences (such as `CSI q`). Shell configurations will often request specific cursor styles.

My guess is that something is requesting for the cursor to become a blinking bar.  Most likely that thing is OrbStack itself, the shell that runs in the Linux machine, or something in your (Linux) shell or OrbStack configuration files.

If you use Fish in your Linux machine, check if the variables `fish_cursor_insert`, `fish_cursor_visual`, `fish_cursor_default`, `fish_cursor_replace`, `fish_cursor_unknown`, `fish_cursor_external`, `fish_cursor_replace_one` or `fish_cursor_selection_mode` are set.

You could also open the terminal inspector (same default keybinding as your web browser's inspector) and, in the “Terminal IO” tab, look for `csi ... q`, where `...` is any **number** or **whitespace**.  That won't narrow it down to which process changes it, but it will tell you *when* it changes.

> I have tried re-sourcing the shell-integration script manually, which does not change anything.

Which is to be expected, as you told it to not touch your cursor with this line:
```ini
shell-integration-features = no-cursor,sudo,title
```

> inside both the OSX terminal and the Linux orb terminal

If you have Ghostty installed in both macOS and Linux, note that the one that matters is the one which *you are currently interacting with*; that means, any configuration or issue on the one that isn't running means nothing in this case.

> As configured, when first launching Ghostty on OSX, it shows a non-blinking block cursor. Then, when entering an OrbStack Linux machine from the terminal, (orb), the cursor becomes a blinking bar.
> 
> The same thing happens when I hit ctrl-d to escape out of Linux and back into the OSX environment. Meaning, the cursor has then become a blinking bar.

If the cursor changed to a blinking bar when you entered Linux, and nothing changed it back, exiting won't reset the cursor unless something in the macOS environment asks it to be reset.

[`cursor-style`]: https://ghostty.org/docs/config/reference#cursor-style

**Comment by daryl-hurst** [comment]
Similar here on MacOS, cursor starts of as a solid white block but after `zsh` has initialised the prompt turns to a single flashing line. So it seems the shell may be affecting behaviour.
Also, if I swap to `bash` within `zsh` the cursor returns back to a solid white block.
Weirdly even if I execute `zsh` whilst within a `zsh` shell the cursor returns to a solid white block.

**Comment by Ragnoroct** [comment]
I believe I have the same problem

When I first start the Ghostty terminal (Ctrl+Alt+T), the cursor appears as a blinking line. Pressing Enter alone doesn't change it, but typing and executing any valid/invalid command updates it to a blinking bar, which is the desired style. The blinking bar is set somewhere in /usr/lib/git-core/git-sh-prompt within __git_ps1. If I don't modify __PROMPT_COMMAND=__git_ps1, the cursor remains a blinking line indefinitely.


```
Ghostty 1.1.0

Version
  - version: 1.1.0
  - channel: stable
Build Config
  - Zig version: 0.13.0
  - build mode : builtin.OptimizeMode.ReleaseFast
  - app runtime: apprt.Runtime.gtk
  - font engine: font.main.Backend.fontconfig_freetype
  - renderer   : renderer.OpenGL
  - libxev     : main.Backend.io_uring
  - desktop env: other
  - GTK version:
    build      : 4.14.2
    runtime    : 4.14.2
  - libadwaita : enabled
    build      : 1.5.0
    runtime    : 1.5.0
  - libX11     : enabled
  - libwayland : enabled
  
GNU bash, version 5.2.21(1)-release (x86_64-pc-linux-gnu)
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

**Reply by 00-kat** [reply]
What's a “blinking bar”, and what's a “blinking line”?

**Reply by Ragnoroct** [reply]
**blinking line**
![image](https://github.com/user-attachments/assets/f38a9ffe-8c0d-4929-b535-aae21ac482f7)

**blinking bar**
![image](https://github.com/user-attachments/assets/577cc47e-fe67-445a-abe6-50d413934307)

**Reply by 00-kat** [reply]
Ah; your issue would be solved with https://github.com/ghostty-org/ghostty/discussions/3893#discussioncomment-12358574 too.

**Comment by V4G4X** [comment]
I'm on MacOS and using Zsh seeing the same issue.  

When my cursor is free to type it is a blinking line, the moment the shell is busy it turns into the proper `block`.

**Comment by 00-kat** [comment, marked as answer]
The issue @yotamNimble, @daryl-hurst, and @V4G4X are facing would be solved with
```ini
shell-integration-features = no-cursor
```

**Reply by V4G4X** [reply]
Thanks that fixed it for me 👍🏽

**Comment by magnusriga** [comment]
@00-kat I just created this perhaps related thread, about a cursor issue: https://github.com/ghostty-org/ghostty/discussions/6091

I already set `shell-integration-features = no-cursor`, so that was not the exact issue at least.

Any thoughts or pointers would be appreciated.
